rust/core-lib/src/selection.rs

```diff
@@ -0,0 +1,309 @@
+// Copyright 2017 Google Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Data structures representing (multiple) selections and cursors.
+
+// TODO: actually use these
+#![allow(unused)]
+
+use std::cmp::{min, max};
+use std::ops::Deref;
+use index_set::remove_n_at;
+
+/// A type representing horizontal measurements. This is currently in units
+/// that are not very well defined except that ASCII characters count as
+/// 1 each. It will change.
+pub type HorizPos = usize;
+
+/// A set of zero or more selection regions, representing a selection state.
+#[derive(Default, Debug)]
+pub struct Selection {
+    // an invariant: regions[i].max() <= regions[i+1].min()
+    // and < if either is_caret()
+    regions: Vec<SelRegion>,
+}
+
+impl Selection {
+    /// Creates a new empty selection.
+    pub fn new() -> Selection {
+        Selection::default()
+    }
+
+    /// Clear the selection.
+    pub fn clear(&mut self) {
+        self.regions.clear();
+    }
+
+    // The smallest index so that offset > region.max() for all preceding
+    // regions.
+    fn search(&self, offset: usize) -> usize {
+        if self.regions.is_empty() || offset > self.regions.last().unwrap().max() {
+            return self.regions.len();
+        }
+        match self.regions.binary_search_by(|r| r.max().cmp(&offset)) {
+            Ok(ix) => ix,
+            Err(ix) => ix,
+        }
+    }
+
+    /// Add a region to the selection. This method implements merging logic.
+    ///
+    /// Two non-caret regions merge if their interiors intersect; merely
```

Comment:
to clarify some terms:

A 'caret region' is a size-zero region, e.g. a cursor? A non-caret region is a size > 0 region, e.g. a selection?